scripts/check_setting_visibility.py

```diff
@@ -0,0 +1,180 @@
+
+
+    def getSettingsFromSettingVisibilityFile(self, file_path: str):
+        parser = configparser.ConfigParser(allow_no_value = True)
+        parser.read([file_path])
+
+        if not parser.has_option("general", "name") or not parser.has_option("general", "weight"):
+            raise NotImplementedError("Visibility setting file missing general data")
+
+        settings = {}
+        for section in parser.sections():
+            if section == 'general':
+                continue
+
+            if section not in settings:
+                settings[section] = []
+
+            for option in parser[section].keys():
+                settings[section].append(option)
+
+        return settings
+
+    def validateSettingsVisibility(self, setting_visibility_items: Dict):
+
+        not_valid_categories = {}
+        not_valid_setting_by_category = {}
+        not_valid_setting_by_order = {}
+
+        visible_settings_order = [] # This list is used to make sure that the settings are in the correct order.
+        # basic.cfg -> advanced.cfg -> expert.cfg. Like: if the setting 'layer_height' in 'basic.cfg' then the same setting
+        # also should be in 'advanced.cfg'
+
+        all_settings_categories = list(self.all_settings_keys.keys())
+
+        # visibility_type = basic, advanced, expert
+        for visibility_type in setting_visibility:
+            item = setting_visibility_items[visibility_type]
+
+            not_valid_setting_by_category[visibility_type] = [] # this list is for keeping invalid settings.
+            not_valid_categories[visibility_type] = []
+
+            for category, category_settings in item.items():
+
+                # Validate Category, If category is not defined then the test will fail
+                if category not in all_settings_categories:
+                    not_valid_categories[visibility_type].append(category)
+
+                for setting in category_settings:
+
+                    # Check whether the setting exist in fdmprinter.def.json or not.
+                    # If the setting is defined in the wrong category or does not exist there then the test will fail
+                    if setting not in self.all_settings_keys[category]:
+                        not_valid_setting_by_category[visibility_type].append(setting)
+
+                    # Add the 'basic' settings to the list
+                    if visibility_type == "basic":
+                        visible_settings_order.append(setting)
+
+
+        # Check whether the settings are added in right order or not.
+        # The basic settings should be in advanced, and advanced in expert
+        for visibility_type in setting_visibility:
+
+            # Skip the basic because it cannot be compared to previous list
+            if visibility_type == 'basic':
+                continue
+
+            all_settings_in_this_type = []
+            not_valid_setting_by_order[visibility_type] = []
```

Comment:
If you set this test up as an automated test, it will be unnecessary to track all of this state. You can simply let it fail on the first setting it encounters is wrong. Correcting it and then re-running the test will reveal the next.